RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitTimingCCAlgo.cc

```diff
@@ -0,0 +1,164 @@
+                                                    const EcalMGPAGainRatio* aGain,
+                                                    const FullSampleVector& fullpulse,
+                                                    EcalUncalibratedRecHit& uncalibRecHit) {
+  const unsigned int nsample = EcalDataFrame::MAXSAMPLES;
+
+  double maxamplitude = -std::numeric_limits<double>::max();
+
+  double pulsenorm = 0.;
+
+  std::vector<double> pedSubSamples(nsample);
+  for (unsigned int iSample = 0; iSample < nsample; iSample++) {
+    const EcalMGPASample& sample = dataFrame.sample(iSample);
+
+    double amplitude = 0.;
+    int gainId = sample.gainId();
+
+    double pedestal = 0.;
+    double gainratio = 1.;
+
+    if (gainId == 0 || gainId == 3) {
+      pedestal = aped->mean_x1;
+      gainratio = aGain->gain6Over1() * aGain->gain12Over6();
+    } else if (gainId == 1) {
+      pedestal = aped->mean_x12;
+      gainratio = 1.;
+    } else if (gainId == 2) {
+      pedestal = aped->mean_x6;
+      gainratio = aGain->gain12Over6();
+    }
+
+    amplitude = ((double)(sample.adc()) - pedestal) * gainratio;
+
+    if (gainId == 0) {
+      //saturation
+      amplitude = (4095. - pedestal) * gainratio;
+    }
+
+    pedSubSamples.at(iSample) = amplitude;
+
+    if (amplitude > maxamplitude) {
+      maxamplitude = amplitude;
+    }
+    pulsenorm += fullpulse(iSample);
+  }
+
+  std::vector<double>::const_iterator amplit;
+  for (amplit = amplitudes.begin(); amplit < amplitudes.end(); ++amplit) {
+    int ipulse = std::distance(amplitudes.begin(), amplit);
+    // The following 3 lines are copied from EcalRecAlgos/interface/EcalUncalibRecHitTimeWeightsAlgo.h
+    int bx = ipulse - 5;
+    int firstsamplet = std::max(0, bx + 3);
+    int offset = 7 - 3 - bx;
+
+    std::vector<double> pulse(nsample);
+    for (unsigned int isample = firstsamplet; isample < nsample; ++isample) {
+      pulse.at(isample) = fullpulse(isample + offset);
+      pedSubSamples.at(isample) =
+          std::max(0., pedSubSamples.at(isample) - amplitudes[ipulse] * pulse.at(isample) / pulsenorm);
+    }
+  }
+
+  float tStart = startTime_ + GLOBAL_TIME_SHIFT;
+  float tStop = stopTime_ + GLOBAL_TIME_SHIFT;
+  float tM = (tStart + tStop) / 2;
+
+  float distStart, distStop;
+  int counter = 0;
+
+  do {
+    ++counter;
+    distStart = computeCC(pedSubSamples, fullpulse, tStart);
+    distStop = computeCC(pedSubSamples, fullpulse, tStop);
+
+    if (distStart > distStop) {
+      tStart = tStart;
+      tStop = tM;
+    } else {
+      tStart = tM;
+      tStop = tStop;
+    }
+    tM = (tStart + tStop) / 2;
+
+  } while (tStop - tStart > TARGET_TIME_PRECISION && counter < MAX_NUM_OF_ITERATIONS);
+
+  tM -= GLOBAL_TIME_SHIFT;
+
+  if (counter < MIN_NUM_OF_ITERATIONS || counter > MAX_NUM_OF_ITERATIONS - 1) {
+    if (counter > MAX_NUM_OF_ITERATIONS / 2)
+      //Produce a log if minimization took too long
+      edm::LogWarning("EcalUncalibRecHitTimingCCAlgo::computeTimeCC")
```

Comment:
How often do you expect this warning to happen?

Reply:
not to often, actually, it never happened to me. But if it happens it is useful to know.
In general, this code was tested for functionality (time precision vs previous algorithms) but not for performance.
The average number of cycles here is ~5 but it could be reduced further and maybe something can be done to improve rare cases where it takes too long (MAX_NUM_OF_ITERATIONS)

Reply:
Is it worth a warning or would logInfo do as well? Your call.

Reply:
I'll leave just a logInfo, since in any case if this happens, the jitterError will be negative (to tell users not to trust the jitter for that rechit)